File: Android App/app/src/main/java/com/justaleks/syncnotes/ai/KeySync.kt
package com.justaleks.syncnotes.ai

import android.util.Base64
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * Optional, off by default: keep an encrypted copy of the API key in the account
 * so a new device can unlock it instead of being handed the key again.
 *
 * The encryption is done with a passphrase the user chooses, and that choice is
 * the whole point. Anything the app could decrypt on its own — a key derived from
 * the uid, a secret shipped in the APK, a value stored beside the ciphertext —
 * would also be decryptable by anyone who reached the account, which is precisely
 * the risk that kept the key out of Firestore in the first place.
 *
 * Deliberately interoperable with the web app: PBKDF2-HMAC-SHA256 at the same
 * iteration count, AES-256-GCM with a 12-byte IV and the 128-bit tag appended to
 * the ciphertext, everything base64. A key encrypted in the browser unlocks here
 * and vice versa — which is the entire point of syncing it.
 */
data class EncryptedKey(
    val provider: AiProvider,
    val model: String,
    val salt: String,
    val iv: String,
    val ct: String,
    val iterations: Int,
)

private const val ITERATIONS = 310_000
private const val KEY_BITS = 256
private const val TAG_BITS = 128
private const val IV_BYTES = 12
private const val SALT_BYTES = 16

/** No line breaks or padding surprises — the web app reads these strings too. */
private const val B64 = Base64.NO_WRAP

object KeySync {

    private val db: FirebaseFirestore get() = FirebaseFirestore.getInstance()

    private fun keyDoc(uid: String) =
        db.collection("users").document(uid).collection("private").document("aiKey")

    private fun deriveKey(passphrase: String, salt: ByteArray, iterations: Int): SecretKeySpec {
        val spec = PBEKeySpec(passphrase.toCharArray(), salt, iterations, KEY_BITS)
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        return SecretKeySpec(factory.generateSecret(spec).encoded, "AES")
    }

    suspend fun encrypt(
        apiKey: String,
        passphrase: String,
        provider: AiProvider,
        model: String,
    ): EncryptedKey = withContext(Dispatchers.Default) {
        val random = SecureRandom()
        val salt = ByteArray(SALT_BYTES).also(random::nextBytes)
        val iv = ByteArray(IV_BYTES).also(random::nextBytes)

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(
            Cipher.ENCRYPT_MODE,
            deriveKey(passphrase, salt, ITERATIONS),
            GCMParameterSpec(TAG_BITS, iv),
        )
        val ct = cipher.doFinal(apiKey.toByteArray(Charsets.UTF_8))

        EncryptedKey(
            provider = provider,
            model = model,
            salt = Base64.encodeToString(salt, B64),
            iv = Base64.encodeToString(iv, B64),
            ct = Base64.encodeToString(ct, B64),
            iterations = ITERATIONS,
        )
    }

    /**
     * Returns null when the passphrase is wrong.
     *
     * AES-GCM authenticates the ciphertext, so a wrong passphrase fails outright
     * rather than yielding plausible rubbish — the user gets told they mistyped
     * instead of quietly ending up with a broken key.
     */
    suspend fun decrypt(blob: EncryptedKey, passphrase: String): String? =
        withContext(Dispatchers.Default) {
            runCatching {
                val cipher = Cipher.getInstance("AES/GCM/NoPadding")
                cipher.init(
                    Cipher.DECRYPT_MODE,
                    deriveKey(passphrase, Base64.decode(blob.salt, B64), blob.iterations),
                    GCMParameterSpec(TAG_BITS, Base64.decode(blob.iv, B64)),
                )
                String(cipher.doFinal(Base64.decode(blob.ct, B64)), Charsets.UTF_8)
            }.getOrNull()
        }

    suspend fun load(uid: String): EncryptedKey? {
        val doc = runCatching { keyDoc(uid).get().await() }.getOrNull() ?: return null
        if (!doc.exists() || doc.getLong("v")?.toInt() != 1) return null

        val salt = doc.getString("salt") ?: return null
        val iv = doc.getString("iv") ?: return null
        val ct = doc.getString("ct") ?: return null

        return EncryptedKey(
            provider = AiProvider.from(doc.getString("provider")),
            model = doc.getString("model").orEmpty(),
            salt = salt,
            iv = iv,
            ct = ct,
            iterations = doc.getLong("iterations")?.toInt() ?: ITERATIONS,
        )
    }

    suspend fun save(uid: String, blob: EncryptedKey) {
        keyDoc(uid).set(
            mapOf(
                "v" to 1,
                "provider" to blob.provider.id,
                "model" to blob.model,
                "salt" to blob.salt,
                "iv" to blob.iv,
                "ct" to blob.ct,
                "iterations" to blob.iterations,
            )
        ).await()
    }

    suspend fun clear(uid: String) {
        keyDoc(uid).delete().await()
    }
}

File: Website/src/ai/keySync.ts
import { deleteDoc, doc, getDoc, setDoc } from 'firebase/firestore'
import { db } from '../firebase'
import type { ProviderId } from './settings'

/**
 * Optional, off by default: keep an encrypted copy of the API key in the account
 * so a new device can unlock it instead of being handed the key again.
 *
 * The encryption is done with a passphrase the user chooses, and that choice is
 * the whole point. Anything the app could decrypt on its own — a key derived from
 * the uid, a secret shipped in the bundle, a value stored beside the ciphertext —
 * would also be decryptable by anyone who reached the account, which is precisely
 * the risk that kept the key out of Firestore in the first place. A passphrase we
 * never store is the only version of this feature that is worth the name.
 *
 * The cost is honest and is spelled out to the user before they turn it on: the
 * passphrase is unrecoverable, and a weak one makes the whole thing weak.
 */
export type EncryptedKey = {
  /** Format version, so this can be changed later without silent breakage. */
  v: 1
  provider: ProviderId
  model: string
  /** Base64. */
  salt: string
  iv: string
  ct: string
  iterations: number
}

/**
 * OWASP's floor for PBKDF2-HMAC-SHA256. Deliberately slow: it is the only thing
 * standing between a weak passphrase and an offline guessing attack on ciphertext
 * that an attacker may already hold.
 */
const ITERATIONS = 310_000

const enc = new TextEncoder()
const dec = new TextDecoder()

const toBase64 = (bytes: ArrayBuffer | Uint8Array) =>
  btoa(String.fromCharCode(...new Uint8Array(bytes)))

const fromBase64 = (text: string) =>
  Uint8Array.from(atob(text), (c) => c.charCodeAt(0))

async function deriveAesKey(
  passphrase: string,
  salt: Uint8Array,
  iterations: number,
): Promise<CryptoKey> {
  const material = await crypto.subtle.importKey(
    'raw',
    enc.encode(passphrase),
    'PBKDF2',
    false,
    ['deriveKey'],
  )

  return crypto.subtle.deriveKey(
    { name: 'PBKDF2', salt: salt as BufferSource, iterations, hash: 'SHA-256' },
    material,
    { name: 'AES-GCM', length: 256 },
    false,
    ['encrypt', 'decrypt'],
  )
}

export async function encryptKey(
  apiKey: string,
  passphrase: string,
  provider: ProviderId,
  model: string,
): Promise<EncryptedKey> {
  const salt = crypto.getRandomValues(new Uint8Array(16))
  const iv = crypto.getRandomValues(new Uint8Array(12))
  const key = await deriveAesKey(passphrase, salt, ITERATIONS)

  const ct = await crypto.subtle.encrypt(
    { name: 'AES-GCM', iv: iv as BufferSource },
    key,
    enc.encode(apiKey),
  )

  return {
    v: 1,
    provider,
    model,
    salt: toBase64(salt),
    iv: toBase64(iv),
    ct: toBase64(ct),
    iterations: ITERATIONS,
  }
}

/**
 * Returns null when the passphrase is wrong.
 *
 * AES-GCM authenticates the ciphertext, so a wrong passphrase fails to decrypt
 * rather than yielding plausible rubbish — the user gets told they mistyped
 * instead of quietly ending up with a broken key.
 */
export async function decryptKey(
  blob: EncryptedKey,
  passphrase: string,
): Promise<string | null> {
  try {
    const key = await deriveAesKey(passphrase, fromBase64(blob.salt), blob.iterations)
    const plain = await crypto.subtle.decrypt(
      { name: 'AES-GCM', iv: fromBase64(blob.iv) as BufferSource },
      key,
      fromBase64(blob.ct) as BufferSource,
    )
    return dec.decode(plain)
  } catch {
    return null
  }
}

/**
 * One document per account, not per device: the point is that every device sees
 * the same envelope.
 */
function keyDoc(uid: string) {
  return doc(db, 'users', uid, 'private', 'aiKey')
}

export async function loadSyncedKey(uid: string): Promise<EncryptedKey | null> {
  const snap = await getDoc(keyDoc(uid))
  if (!snap.exists()) return null

  const data = snap.data() as Partial<EncryptedKey>
  if (data.v !== 1 || !data.salt || !data.iv || !data.ct) return null

  return {
    v: 1,
    provider: data.provider === 'openai' ? 'openai' : 'anthropic',
    model: typeof data.model === 'string' ? data.model : '',
    salt: data.salt,
    iv: data.iv,
    ct: data.ct,
    iterations: typeof data.iterations === 'number' ? data.iterations : ITERATIONS,
  }
}

export async function saveSyncedKey(uid: string, blob: EncryptedKey): Promise<void> {
  await setDoc(keyDoc(uid), blob)
}

export async function clearSyncedKey(uid: string): Promise<void> {
  await deleteDoc(keyDoc(uid))
}
